Balance node usage before creating victim pods in preemption e2e

damemi opened the pull request:
**What type of PR is this?**
/kind failing-test

**What this PR does / why we need it**:
These preemption e2es attempt to create filler "victim" pods that use 60% of a node's available resources, and assumes that with all nodes full, the low priority pod will get evicted.

However, if the nodes have very different usages (say one is at 500/1500 cpu and another is at 1000/1500), then the request of 60% of the low-priority-pod's-node's usage may not fit on one node, but will fit on another and the test fails.

This PR creates balanced pods in these e2es to ensure a level field for the preemption tests.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

damemi committed:
Balance node usage before creating victim pods in preemption e2e

damemi commented:
/hold
continuing to debug these tests, it seems they might have more problems than just this

damemi commented:
I think `createBalancedPodForNodes` is either broken or its description is inaccurate, I expect it to make all nodes have the same mem/usage ratio but in these logs I see that doesn't happen:
```
STEP: Create pods that use 60% of node resources.
May  4 22:24:59.072: INFO: ComputeCPUMemFraction for node: ip-10-0-136-231.us-east-2.compute.internal
May  4 22:24:59.189: INFO: Pod for on the node: tuned-5qsth, Cpu: 10, Mem: 52428800
May  4 22:24:59.189: INFO: Pod for on the node: csi-snapshot-controller-7c8b5c7948-8bjdz, Cpu: 10, Mem: 52428800
May  4 22:24:59.189: INFO: Pod for on the node: dns-default-6n8vz, Cpu: 65, Mem: 325058560
May  4 22:24:59.189: INFO: Pod for on the node: image-registry-6cd6657449-vq9vk, Cpu: 100, Mem: 268435456
May  4 22:24:59.189: INFO: Pod for on the node: node-ca-56nv8, Cpu: 10, Mem: 10485760
May  4 22:24:59.189: INFO: Pod for on the node: router-default-5b68bd5df6-m9m9d, Cpu: 100, Mem: 268435456
May  4 22:24:59.189: INFO: Pod for on the node: migrator-5b6cfb6b66-7j8fb, Cpu: 100, Mem: 209715200
May  4 22:24:59.189: INFO: Pod for on the node: machine-config-daemon-wh9qw, Cpu: 40, Mem: 104857600
May  4 22:24:59.189: INFO: Pod for on the node: alertmanager-main-0, Cpu: 8, Mem: 482344960
May  4 22:24:59.189: INFO: Pod for on the node: alertmanager-main-2, Cpu: 8, Mem: 482344960
May  4 22:24:59.189: INFO: Pod for on the node: node-exporter-xr2jx, Cpu: 9, Mem: 220200960
May  4 22:24:59.189: INFO: Pod for on the node: prometheus-adapter-7d86fd7c5f-sllr5, Cpu: 1, Mem: 26214400
May  4 22:24:59.189: INFO: Pod for on the node: prometheus-k8s-0, Cpu: 76, Mem: 1660944384
May  4 22:24:59.189: INFO: Pod for on the node: thanos-querier-cb8dd48b4-lrnxj, Cpu: 8, Mem: 75497472
May  4 22:24:59.189: INFO: Pod for on the node: multus-tx8qf, Cpu: 10, Mem: 157286400
May  4 22:24:59.189: INFO: Pod for on the node: ovs-47ccm, Cpu: 100, Mem: 419430400
May  4 22:24:59.189: INFO: Pod for on the node: sdn-vx5mp, Cpu: 100, Mem: 209715200
May  4 22:24:59.189: INFO: Node: ip-10-0-136-231.us-east-2.compute.internal, totalRequestedCPUResource: 855, cpuAllocatableMil: 3500, cpuFraction: 0.24428571428571427
May  4 22:24:59.189: INFO: Node: ip-10-0-136-231.us-east-2.compute.internal, totalRequestedMemResource: 5130682368, memAllocatableVal: 15634849792, memFraction: 0.32815680587000295
May  4 22:24:59.189: INFO: ComputeCPUMemFraction for node: ip-10-0-139-24.us-east-2.compute.internal
May  4 22:24:59.273: INFO: Pod for on the node: tuned-jm66k, Cpu: 10, Mem: 52428800
May  4 22:24:59.273: INFO: Pod for on the node: dns-default-jzf5g, Cpu: 65, Mem: 325058560
May  4 22:24:59.273: INFO: Pod for on the node: node-ca-w6klx, Cpu: 10, Mem: 10485760
May  4 22:24:59.273: INFO: Pod for on the node: machine-config-daemon-4cgrj, Cpu: 40, Mem: 104857600
May  4 22:24:59.273: INFO: Pod for on the node: node-exporter-m8crj, Cpu: 9, Mem: 220200960
May  4 22:24:59.273: INFO: Pod for on the node: multus-pdh86, Cpu: 10, Mem: 157286400
May  4 22:24:59.273: INFO: Pod for on the node: ovs-f2k79, Cpu: 100, Mem: 419430400
May  4 22:24:59.273: INFO: Pod for on the node: sdn-swkdl, Cpu: 100, Mem: 209715200
May  4 22:24:59.273: INFO: Node: ip-10-0-139-24.us-east-2.compute.internal, totalRequestedCPUResource: 444, cpuAllocatableMil: 3500, cpuFraction: 0.12685714285714286
May  4 22:24:59.273: INFO: Node: ip-10-0-139-24.us-east-2.compute.internal, totalRequestedMemResource: 1604321280, memAllocatableVal: 15634849792, memFraction: 0.10261187675886052
May  4 22:24:59.273: INFO: ComputeCPUMemFraction for node: ip-10-0-150-85.us-east-2.compute.internal
May  4 22:24:59.368: INFO: Pod for on the node: tuned-m9b96, Cpu: 10, Mem: 52428800
May  4 22:24:59.368: INFO: Pod for on the node: csi-snapshot-controller-operator-877994c6c-nvlcq, Cpu: 10, Mem: 52428800
May  4 22:24:59.368: INFO: Pod for on the node: dns-default-dsvqx, Cpu: 65, Mem: 325058560
May  4 22:24:59.368: INFO: Pod for on the node: image-registry-6cd6657449-cs8hm, Cpu: 100, Mem: 268435456
May  4 22:24:59.368: INFO: Pod for on the node: node-ca-7fm6c, Cpu: 10, Mem: 10485760
May  4 22:24:59.368: INFO: Pod for on the node: router-default-5b68bd5df6-z7vnb, Cpu: 100, Mem: 268435456
May  4 22:24:59.368: INFO: Pod for on the node: machine-config-daemon-j9dcq, Cpu: 40, Mem: 104857600
May  4 22:24:59.368: INFO: Pod for on the node: certified-operators-5bcfb9dc96-wqhc7, Cpu: 10, Mem: 104857600
May  4 22:24:59.368: INFO: Pod for on the node: community-operators-c9cf4bd9d-vdgmz, Cpu: 10, Mem: 104857600
May  4 22:24:59.368: INFO: Pod for on the node: redhat-marketplace-555748dff5-r42w4, Cpu: 10, Mem: 104857600
May  4 22:24:59.368: INFO: Pod for on the node: redhat-operators-6986b76dff-sp9tm, Cpu: 10, Mem: 104857600
May  4 22:24:59.368: INFO: Pod for on the node: alertmanager-main-1, Cpu: 8, Mem: 482344960
May  4 22:24:59.368: INFO: Pod for on the node: grafana-bb94c59bd-jlxzw, Cpu: 5, Mem: 125829120
May  4 22:24:59.368: INFO: Pod for on the node: kube-state-metrics-7b97875b8b-mqx56, Cpu: 4, Mem: 125829120
May  4 22:24:59.368: INFO: Pod for on the node: node-exporter-4xpm8, Cpu: 9, Mem: 220200960
May  4 22:24:59.368: INFO: Pod for on the node: openshift-state-metrics-dcb7dddb4-wq8ch, Cpu: 3, Mem: 199229440
May  4 22:24:59.368: INFO: Pod for on the node: prometheus-adapter-7d86fd7c5f-nrnw7, Cpu: 1, Mem: 26214400
May  4 22:24:59.368: INFO: Pod for on the node: prometheus-k8s-1, Cpu: 76, Mem: 1660944384
May  4 22:24:59.368: INFO: Pod for on the node: telemeter-client-664c494db4-ls2xh, Cpu: 3, Mem: 440401920
May  4 22:24:59.368: INFO: Pod for on the node: thanos-querier-cb8dd48b4-x9wtn, Cpu: 8, Mem: 75497472
May  4 22:24:59.368: INFO: Pod for on the node: multus-cfcfx, Cpu: 10, Mem: 157286400
May  4 22:24:59.368: INFO: Pod for on the node: ovs-wpkqw, Cpu: 100, Mem: 419430400
May  4 22:24:59.368: INFO: Pod for on the node: sdn-kkg56, Cpu: 100, Mem: 209715200
May  4 22:24:59.368: INFO: Node: ip-10-0-150-85.us-east-2.compute.internal, totalRequestedCPUResource: 802, cpuAllocatableMil: 3500, cpuFraction: 0.22914285714285715
May  4 22:24:59.368: INFO: Node: ip-10-0-150-85.us-east-2.compute.internal, totalRequestedMemResource: 5749342208, memAllocatableVal: 15634849792, memFraction: 0.3677260916789753
May  4 22:24:59.414: INFO: Waiting for running...
May  4 22:25:04.501: INFO: Waiting for running...
May  4 22:25:09.589: INFO: Waiting for running...
STEP: Compute Cpu, Mem Fraction after create balanced pods.
May  4 22:25:14.639: INFO: ComputeCPUMemFraction for node: ip-10-0-136-231.us-east-2.compute.internal
May  4 22:25:14.776: INFO: Pod for on the node: 1554b43b-cff3-4734-9fa1-06932276fde8-0, Cpu: 1655, Mem: 7776588595
May  4 22:25:14.776: INFO: Pod for on the node: tuned-5qsth, Cpu: 10, Mem: 52428800
May  4 22:25:14.776: INFO: Pod for on the node: csi-snapshot-controller-7c8b5c7948-8bjdz, Cpu: 10, Mem: 52428800
May  4 22:25:14.776: INFO: Pod for on the node: dns-default-6n8vz, Cpu: 65, Mem: 325058560
May  4 22:25:14.776: INFO: Pod for on the node: image-registry-6cd6657449-vq9vk, Cpu: 100, Mem: 268435456
May  4 22:25:14.776: INFO: Pod for on the node: node-ca-56nv8, Cpu: 10, Mem: 10485760
May  4 22:25:14.776: INFO: Pod for on the node: router-default-5b68bd5df6-m9m9d, Cpu: 100, Mem: 268435456
May  4 22:25:14.776: INFO: Pod for on the node: migrator-5b6cfb6b66-7j8fb, Cpu: 100, Mem: 209715200
May  4 22:25:14.776: INFO: Pod for on the node: machine-config-daemon-wh9qw, Cpu: 40, Mem: 104857600
May  4 22:25:14.776: INFO: Pod for on the node: alertmanager-main-0, Cpu: 8, Mem: 482344960
May  4 22:25:14.776: INFO: Pod for on the node: alertmanager-main-2, Cpu: 8, Mem: 482344960
May  4 22:25:14.776: INFO: Pod for on the node: node-exporter-xr2jx, Cpu: 9, Mem: 220200960
May  4 22:25:14.776: INFO: Pod for on the node: prometheus-adapter-7d86fd7c5f-sllr5, Cpu: 1, Mem: 26214400
May  4 22:25:14.776: INFO: Pod for on the node: prometheus-k8s-0, Cpu: 76, Mem: 1660944384
May  4 22:25:14.776: INFO: Pod for on the node: thanos-querier-cb8dd48b4-lrnxj, Cpu: 8, Mem: 75497472
May  4 22:25:14.776: INFO: Pod for on the node: multus-tx8qf, Cpu: 10, Mem: 157286400
May  4 22:25:14.776: INFO: Pod for on the node: ovs-47ccm, Cpu: 100, Mem: 419430400
May  4 22:25:14.776: INFO: Pod for on the node: sdn-vx5mp, Cpu: 100, Mem: 209715200
May  4 22:25:14.776: INFO: Node: ip-10-0-136-231.us-east-2.compute.internal, totalRequestedCPUResource: 2510, cpuAllocatableMil: 3500, cpuFraction: 0.7171428571428572
May  4 22:25:14.776: INFO: Node: ip-10-0-136-231.us-east-2.compute.internal, totalRequestedMemResource: 12907270963, memAllocatableVal: 15634849792, memFraction: 0.8255449290983505
STEP: Compute Cpu, Mem Fraction after create balanced pods.
May  4 22:25:14.776: INFO: ComputeCPUMemFraction for node: ip-10-0-139-24.us-east-2.compute.internal
May  4 22:25:14.868: INFO: Pod for on the node: 184f69a8-05a9-4dc8-b960-ba6f875b3096-0, Cpu: 1245, Mem: 4250227507
May  4 22:25:14.868: INFO: Pod for on the node: tuned-jm66k, Cpu: 10, Mem: 52428800
May  4 22:25:14.868: INFO: Pod for on the node: dns-default-jzf5g, Cpu: 65, Mem: 325058560
May  4 22:25:14.868: INFO: Pod for on the node: node-ca-w6klx, Cpu: 10, Mem: 10485760
May  4 22:25:14.868: INFO: Pod for on the node: machine-config-daemon-4cgrj, Cpu: 40, Mem: 104857600
May  4 22:25:14.868: INFO: Pod for on the node: node-exporter-m8crj, Cpu: 9, Mem: 220200960
May  4 22:25:14.868: INFO: Pod for on the node: multus-pdh86, Cpu: 10, Mem: 157286400
May  4 22:25:14.868: INFO: Pod for on the node: ovs-f2k79, Cpu: 100, Mem: 419430400
May  4 22:25:14.868: INFO: Pod for on the node: sdn-swkdl, Cpu: 100, Mem: 209715200
May  4 22:25:14.868: INFO: Node: ip-10-0-139-24.us-east-2.compute.internal, totalRequestedCPUResource: 1689, cpuAllocatableMil: 3500, cpuFraction: 0.4825714285714286
May  4 22:25:14.868: INFO: Node: ip-10-0-139-24.us-east-2.compute.internal, totalRequestedMemResource: 5854548787, memAllocatableVal: 15634849792, memFraction: 0.3744550708760656
STEP: Compute Cpu, Mem Fraction after create balanced pods.
May  4 22:25:14.868: INFO: ComputeCPUMemFraction for node: ip-10-0-150-85.us-east-2.compute.internal
May  4 22:25:14.963: INFO: Pod for on the node: 3f008232-73db-40de-889c-8b1177854baa-0, Cpu: 1298, Mem: 3631567667
May  4 22:25:14.963: INFO: Pod for on the node: tuned-m9b96, Cpu: 10, Mem: 52428800
May  4 22:25:14.963: INFO: Pod for on the node: csi-snapshot-controller-operator-877994c6c-nvlcq, Cpu: 10, Mem: 52428800
May  4 22:25:14.963: INFO: Pod for on the node: dns-default-dsvqx, Cpu: 65, Mem: 325058560
May  4 22:25:14.963: INFO: Pod for on the node: image-registry-6cd6657449-cs8hm, Cpu: 100, Mem: 268435456
May  4 22:25:14.963: INFO: Pod for on the node: node-ca-7fm6c, Cpu: 10, Mem: 10485760
May  4 22:25:14.963: INFO: Pod for on the node: router-default-5b68bd5df6-z7vnb, Cpu: 100, Mem: 268435456
May  4 22:25:14.963: INFO: Pod for on the node: machine-config-daemon-j9dcq, Cpu: 40, Mem: 104857600
May  4 22:25:14.963: INFO: Pod for on the node: certified-operators-5bcfb9dc96-wqhc7, Cpu: 10, Mem: 104857600
May  4 22:25:14.963: INFO: Pod for on the node: community-operators-c9cf4bd9d-vdgmz, Cpu: 10, Mem: 104857600
May  4 22:25:14.963: INFO: Pod for on the node: redhat-marketplace-555748dff5-r42w4, Cpu: 10, Mem: 104857600
May  4 22:25:14.963: INFO: Pod for on the node: redhat-operators-6986b76dff-sp9tm, Cpu: 10, Mem: 104857600
May  4 22:25:14.963: INFO: Pod for on the node: alertmanager-main-1, Cpu: 8, Mem: 482344960
May  4 22:25:14.963: INFO: Pod for on the node: grafana-bb94c59bd-jlxzw, Cpu: 5, Mem: 125829120
May  4 22:25:14.963: INFO: Pod for on the node: kube-state-metrics-7b97875b8b-mqx56, Cpu: 4, Mem: 125829120
May  4 22:25:14.963: INFO: Pod for on the node: node-exporter-4xpm8, Cpu: 9, Mem: 220200960
May  4 22:25:14.963: INFO: Pod for on the node: openshift-state-metrics-dcb7dddb4-wq8ch, Cpu: 3, Mem: 199229440
May  4 22:25:14.963: INFO: Pod for on the node: prometheus-adapter-7d86fd7c5f-nrnw7, Cpu: 1, Mem: 26214400
May  4 22:25:14.964: INFO: Pod for on the node: prometheus-k8s-1, Cpu: 76, Mem: 1660944384
May  4 22:25:14.964: INFO: Pod for on the node: telemeter-client-664c494db4-ls2xh, Cpu: 3, Mem: 440401920
May  4 22:25:14.964: INFO: Pod for on the node: thanos-querier-cb8dd48b4-x9wtn, Cpu: 8, Mem: 75497472
May  4 22:25:14.964: INFO: Pod for on the node: multus-cfcfx, Cpu: 10, Mem: 157286400
May  4 22:25:14.964: INFO: Pod for on the node: ovs-wpkqw, Cpu: 100, Mem: 419430400
May  4 22:25:14.964: INFO: Pod for on the node: sdn-kkg56, Cpu: 100, Mem: 209715200
May  4 22:25:14.964: INFO: Node: ip-10-0-150-85.us-east-2.compute.internal, totalRequestedCPUResource: 2100, cpuAllocatableMil: 3500, cpuFraction: 0.6
May  4 22:25:14.964: INFO: Node: ip-10-0-150-85.us-east-2.compute.internal, totalRequestedMemResource: 9380909875, memAllocatableVal: 15634849792, memFraction: 0.5999999999872081
```
Where I've specified `60%` but end up with `.48`, `.6` and `.72` cpu ratios

oomichi commented on test/e2e/scheduling/priorities.go:
```diff
@@ -464,17 +464,19 @@ func createBalancedPodForNodes(f *framework.Framework, cs clientset.Interface, n
+		if podConfig == nil {
+			podConfig = &pausePodConfig{
 				Name:   "",
 				Labels: balancePodLabel,
```
Need to fix indents to pass gofmt check.

oomichi reviewed (commented):
The change itself seems good for me except small points.

> I think createBalancedPodForNodes is either broken or its description is inaccurate, I expect it to make all nodes have the same mem/usage ratio but in these logs I see that doesn't happen:

The last one(ip-10-0-150-85.us-east-2.compute.internal) only seems to have accurate resource usage.

/cc @oomichi

damemi commented:
I believe at this point I have found the errors with this test and have made the following changes in a new commit:

* `createBalancedPodForResources` attempts to make each node have the same amount of usage by looking at the highest fraction and then topping up every other node to match that. However when it creates the balancing pods it does not ensure that these pods end up on the nodes they are supposed to, which is why I was getting "balanced" fractions that were all over the place in https://github.com/kubernetes/kubernetes/pull/90740#issuecomment-624057909, because the pods could go on literally any node (sometimes several on one node). Added a `NodeAffinity` to the balancing pods to fix this

* Similar to that, the low/medium "victim" pods are supposed to be scheduled one to each node in a loop, but there was nothing to ensure they actually landed on the expected nodes. Added a `NodeAffinity` to fix that (and removed the `NodeName` field, as this does nothing)

* In the calls to `createBalancedPodForNodes` and `getCurrentPodUsageOnTheNode` we were passing `podRequestedResource`, which is a resource request specifying `100Mi` memory and `100m` cpu. This request only had the effect of adding `100` to our calculations which threw off the balancing for no reason. I assume this field is meant to balance a node when you already have a specific request in mind, so that it can be factored in. In our case, we don't, so instead set this to 0.

* Increased the victim request ratios to `70%` of the node usage (up from 60). Need this because the current node usages we calculate in the tests end up higher than the requests that are checked by the NodeResources Fit plugin. I don't know why these differ, but it meant that in the test we were assuming nodes are more full than they are (at least from the scheduler's perspective) and then preemption would not actually need to occur during scheduling of the `preemptor-pod`. I imagine this may still occasionally flake until we sort out and standardize usage reporting, but increasing the ratio helps make sure the nodes are actually full enough to trigger preemption.

@ahg-g or @Huang-Wei would you ptal at this?

damemi commented:
/retest

Huang-Wei commented:
@damemi I'm not a fan of `createBalancedPodForResources` in any way - I feel like it can raise flakes at any time.

Do you think we can come up with a thorough solution to deprecate `createBalancedPodForResources`? Such as using fake extended resource for each e2e which is now relying on `createBalancedPodForResources`.

damemi commented:
/retest

damemi commented:
@Huang-Wei that's a great idea, I think that because these preemption e2e's are fundamentally based on filtering nodes by resource usage that extended resources can definitely eliminate the need for `createBalancedPodForNodes` here.

However, the other use cases of that function are in priorities e2es, where the node balancing is more of a side effect to ensure that resourcebalancing priorities don't interfere with the scheduling. So those are tougher to deprecate `createBalancedPodForNodes` because we do still need to try to balance memory and cpu on the nodes. Personally though, I think the priority e2es themselves are prone to flaking by the nature of how scoring works, and wonder if these could be effectively covered by integration tests.

But I think addressing the other tests' usage of that function is a different PR. I made the change to extended resources for preemption (in a new commit for now) and reverted some of my other changes (but kept the affinity in `createBalancedPodForNodes` which will help the other tests until it/they can be deprecated).

damemi commented:
/retest

Huang-Wei commented:
> Personally though, I think the priority e2es themselves are prone to flaking by the nature of how scoring works, and wonder if these could be effectively covered by integration tests.

I felt the same :(

> But I think addressing the other tests' usage of that function is a different PR.

Sure. How about leveraging multiple-profiles? For instance, spin up the scheduler with N profiles, with each profile corresponding to each Priority (suppose we're now testing one Priority each time), and in that profile set its weight to a big number (e.g., 1000). And in runtime, for each e2e test, specify the Pod to use a particular profile. In this way, the Priority we want to test is less likely to flake as its Score won't be easily diluted.

Huang-Wei commented on test/e2e/scheduling/preemption.go:
```diff
@@ -103,30 +102,20 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
-				continue
-			}
+			// Update each node to advertise 3 available extended resources
+			node.Status.Capacity["test.com/foo"] = resource.MustParse("3")
```
Can you clean up this resource in `ginkgo.AfterEach`? and extract the name "test.com/foo" to an upper variable would also be favorable.

damemi replied:
yes, I also switched to a better resource name `scheduling.k8s.io/foo`

Huang-Wei commented on test/e2e/scheduling/preemption.go:
```diff
@@ -138,8 +127,19 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
+					NodeAffinity: &v1.NodeAffinity{
+						RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+							NodeSelectorTerms: []v1.NodeSelectorTerm{
+								{MatchExpressions: []v1.NodeSelectorRequirement{
+									{Key: "kubernetes.io/hostname", Operator: v1.NodeSelectorOpIn, Values: []string{node.Labels["kubernetes.io/hostname"]}},
```
It'd be more robust to replace this with `MatchFields`, as the node may not have "kubernetes.io/hostname" label: (applies elsewhere)

```go
MatchFields: []v1.NodeSelectorRequirement{
    {Key: "metadata.name", Operator: v1.NodeSelectorOpNotIn, Values: []string{node.Name}},
},
```

damemi replied:
> v1.NodeSelectorOpNotIn

I'm assuming you mean OpIn, updated

Huang-Wei replied:
ah, yes..

Huang-Wei commented on test/e2e/scheduling/preemption.go:
```diff
@@ -162,8 +162,8 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
 			PriorityClassName: highPriorityClassName,
 			Resources: &v1.ResourceRequirements{
 				Requests: podRes,
+				Limits:   podRes,
```
Do we need `Limits`?

damemi replied:
Yeah, without them I got an error that non-over-committable resources need to have a limit set

Huang-Wei reviewed (commented):
Some comments below.

damemi commented:
@Huang-Wei updated with your feedback and squashed, thanks

Huang-Wei commented on test/e2e/scheduling/preemption.go:
```diff
@@ -75,6 +76,10 @@ var _ = SIGDescribe("SchedulerPreemption [Serial]", func() {
 		for _, pair := range priorityPairs {
 			cs.SchedulingV1().PriorityClasses().Delete(context.TODO(), pair.name, *metav1.NewDeleteOptions(0))
 		}
+		for _, node := range nodeList.Items{
```
```suggestion
		for _, node := range nodeList.Items {
```

damemi replied:
Ah yeah, this would fail gofmt check anyway I think

Huang-Wei commented on test/e2e/scheduling/priorities.go:
```diff
@@ -465,16 +465,28 @@ func createBalancedPodForNodes(f *framework.Framework, cs clientset.Interface, n
+					RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+						NodeSelectorTerms: []v1.NodeSelectorTerm{
+							{MatchExpressions: []v1.NodeSelectorRequirement{
+								{Key: "kubernetes.io/hostname", Operator: v1.NodeSelectorOpIn, Values: []string{node.Labels["kubernetes.io/hostname"]}},
```
Please update this with `MatchFields` as well.

damemi replied:
done

Huang-Wei reviewed (commented):
Just some minor comments. LGTM otherwise.

Huang-Wei commented:
Thanks @damemi !

/lgtm
/approve

damemi commented:
/retest

denkensk commented:
/retest

Huang-Wei commented:
/hold cancel
